File: topics/topics/binarysearch/PeakElement.java
package binarysearch;

public class PeakElement {
    public int findPeakElement(int[] nums) {
        //m > l && m > h -> answer -> return idx.
        //m > l && m < h -> ascending -> move right.
        //m < l && m > h -> descending -> move left.
        //m < l && m < h -> can do both -> chosing right for now.

        int low = 0;
        int high = nums.length - 1;

        while (low <= high) {
            int mid = low + (high - low) / 2;

            boolean greaterThanLeft = mid == 0 ? true : nums[mid] > nums[mid - 1];
            boolean greaterThanRight = mid == nums.length - 1 ? true : nums[mid] > nums[mid + 1];

            if (greaterThanLeft && greaterThanRight) {
                return mid;
            }

            if (greaterThanLeft && !greaterThanRight) {
                low = mid + 1;
            }
            // else if(!greaterThanLeft && greaterThanRight){
            else {
                high = mid - 1;
            }

        }

        return -1;
    }

    public static void main(String[] args) {

        PeakElement peakElement = new PeakElement();
        peakElement.findPeakElement(new int[]{1, 2, 1, 3, 5, 6, 4});
    }
}

File: topics/topics/binarysearch/peak_element.py
class Solution:
    def findPeakElement(self, nums: List[int]) -> int:
        s = 0
        e = len(nums) - 1
        if(len(nums)==1):
            return 0


        return self.findPeak(nums, s, e)

    def findPeak(self, nums: List[int], s, e) -> int:
        while s <= e:
            m = s + (e-s)//2
            print(f's={s}, m={m}, e={e}')

            if nums[m-1] <= nums[m] and nums[m] >= nums[(m+1)%len(nums)]:
                return m
            greater = 'left' if nums[m-1] > nums[(m+1)%len(nums)] else 'right'
            if greater == 'right': # move right
                return self.findPeak(nums, m+1, e)
            else:
                return self.findPeak(nums, s, m -1)
        return -1
